pkg/ttl/cache/table.go

```diff
@@ -287,15 +288,31 @@ func (t *PhysicalTable) SplitScanRanges(ctx context.Context, store kv.Storage, s
 	switch ft.GetType() {
 	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeInt24:
 		if len(t.KeyColumns) > 1 {
-			return t.splitCommonHandleRanges(ctx, tikvStore, splitCnt, true, mysql.HasUnsignedFlag(ft.GetFlag()))
+			return t.splitCommonHandleRanges(ctx, tikvStore, splitCnt, true, mysql.HasUnsignedFlag(ft.GetFlag()), nil)
 		}
 		return t.splitIntRanges(ctx, tikvStore, splitCnt)
 	case mysql.TypeBit:
-		return t.splitCommonHandleRanges(ctx, tikvStore, splitCnt, false, false)
+		return t.splitCommonHandleRanges(ctx, tikvStore, splitCnt, false, false, nil)
 	case mysql.TypeString, mysql.TypeVarString, mysql.TypeVarchar:
-		if mysql.HasBinaryFlag(ft.GetFlag()) {
-			return t.splitCommonHandleRanges(ctx, tikvStore, splitCnt, false, false)
+		var decode func([]byte) types.Datum
+		if !mysql.HasBinaryFlag(ft.GetFlag()) {
+			switch ft.GetCharset() {
+			case charset.CharsetASCII, charset.CharsetLatin1:
+				// ASCII and Latin1 are 8-bit charset, we can use GetASCIIPrefixDatumFromBytes to decode it.
+				decode = GetASCIIPrefixDatumFromBytes
+			case charset.CharsetUTF8, charset.CharsetUTF8MB4:
+				switch ft.GetCollate() {
+				case charset.CollationUTF8, charset.CollationUTF8MB4:
```

Comment:
I don't think collations matter here: they don't change the encoding of record keys.